contracts/eip721/EIP721.sol

```diff
@@ -0,0 +1,311 @@
+
+    mapping(uint256 => string) internal tokenURIs;
+
+    // base ERC721 interface = 0x80ac58cd
+    // metadata interface = 0x5b5e139f
+    // enumerable interface = 0x780e9d63
+    bytes4 internal constant ERC721_BASE_INTERFACE_SIGNATURE = 0x80ac58cd;
+    bytes4 internal constant ERC721_METADATA_INTERFACE_SIGNATURE = 0x5b5e139f;
+    bytes4 internal constant ERC721_ENUMERABLE_INTERFACE_SIGNATURE = 0x780e9d63;
+
+    /* Modifiers */
+    modifier tokenExists(uint256 _tokenId) {
+        require(ownerOfToken[_tokenId] != 0);
+        _;
+    }
+
+    // checks: is the owner of the token == msg.sender?
+    // OR has the owner of the token granted msg.sender access to operate?
+    modifier allowedToOperate(uint256 _tokenId) {
+        require(checkIfAllowedToOperate(_tokenId));
+        _;
+    }
+
+    modifier allowedToTransfer(address _from, address _to, uint256 _tokenId) {
+        require(checkIfAllowedToOperate(_tokenId) || approvedOwnerOfToken[_tokenId] == msg.sender);
+        require(ownerOfToken[_tokenId] == _from);
+        require(_to != 0); //not allowed to burn in transfer method
+        _;
+    }
+
+    /// @notice Transfer ownership of an NFT -- THE CALLER IS RESPONSIBLE
+    ///  TO CONFIRM THAT `_to` IS CAPABLE OF RECEIVING NFTS OR ELSE
+    ///  THEY MAY BE PERMANENTLY LOST
+    /// @dev Throws unless `msg.sender` is the current owner, an authorized
+    ///  operator, or the approved address for this NFT. Throws if `_from` is
+    ///  not the current owner. Throws if `_to` is the zero address. Throws if
+    ///  `_tokenId` is not a valid NFT.
+    /// @param _from The current owner of the NFT
+    /// @param _to The new owner
+    /// @param _tokenId The NFT to transfer
+    function transferFrom(address _from, address _to, uint256 _tokenId) external payable
+    tokenExists(_tokenId)
+    allowedToTransfer(_from, _to, _tokenId) {
+        //transfer token
+        settleTransfer(_from, _to, _tokenId);
+    }
+
+    /// @notice Transfers the ownership of an NFT from one address to another address
+    /// @dev Throws unless `msg.sender` is the current owner, an authorized
+    ///  operator, or the approved address for this NFT. Throws if `_from` is
+    ///  not the current owner. Throws if `_to` is the zero address. Throws if
+    ///  `_tokenId` is not a valid NFT. When transfer is complete, this function
+    ///  checks if `_to` is a smart contract (code size > 0). If so, it calls
+    ///  `onERC721Received` on `_to` and throws if the return value is not
+    ///  `bytes4(keccak256("onERC721Received(address,uint256,bytes)"))`.
+    /// @param _from The current owner of the NFT
+    /// @param _to The new owner
+    /// @param _tokenId The NFT to transfer
+    /// @param data Additional data with no specified format, sent in call to `_to`
+    function safeTransferFrom(address _from, address _to, uint256 _tokenId, bytes data) external payable
```

Comment:
Regarding naming: I believe `data` can be `_data` (or the other parameters can have no underscore), since the function ID only uses the types of the parameters, not their Solidity names.